Add new icons, font templates, statistics naming changes and colour updates

peppy opened the pull request:
These are all extracted from song select changes made by @frenzibyte. The goal is to get review eyes on pieces which have scope larger than song select v2 itself, and to reduce the number of external files touched by the main changes.

Of note, I am interested in others' opinions on the font styles. Once this PR is approved, I would be looking to replace all existing usages with the template versions (as a follow-up change).

frenzibyte committed:
Add `DarkOrange` colour set

frenzibyte committed:
Add specifications for new song select icons

frenzibyte committed:
Shorten beatmap hit statistics names

frenzibyte committed:
Define standard font sizes

peppy committed:
Update difficulty spectrum retrieval function

peppy commented:
Okay this should be ready for moving forward with. I've updated a subset of font usage to the new terminology to give an idea of how much coverage we get with the new specifications.

A future step would be to remove all usage of `GetFont()` (there's hundreds more of these which can be updated to `.Style.Body` or equivalent) but I don't want to aim for 100% coverage in this PR as that feels like wasted time right now.

To be clear: I updated the subset to validate the new size and weight combinations are common enough to use as initial template values. If it helps merge this change, the last commit can be dropped and applied separately.

bdach reviewed (commented):
I refuse to check the last commit of this PR, and disclaim myself from any mistakes within or consequences thereof

bdach commented on osu.Game/Graphics/OsuColour.cs:
```diff
@@ -38,7 +34,19 @@ public Color4 ForStarDifficulty(double starDifficulty) => ColourUtils.SampleFrom
+        /// </summary>
+        public Color4 ForStarDifficulty(double starDifficulty, bool showGrayOnZero = true)
+        {
+            if (showGrayOnZero && starDifficulty < 0.1f)
```
I guess to offer something more productive: what is the flag for? where is it being turned off? *does it* need to be turned off?

is this just for the rainbow star rating slider or whatever? if it is can't you use `STAR_DIFFICULTY_SPECTRUM` directly there or something?

frenzibyte replied:
It is for the star rating range slider yes, if I were to use `STAR_DIFFICULTY_SPECTRUM` directly with the gray colour inside there I would have to trim it down from the beginning and readjust, it wouldn't be pretty code.

I'll just apply rounding before the check to preserve behaviour, good note.

frenzibyte replied:
has been removed in https://github.com/ppy/osu/commit/08c17bdf9ec27f9ee8ab35f9c9c2a3ae93b517f3

bdach commented on osu.Game/Graphics/OsuColour.cs:
```diff
@@ -38,7 +34,19 @@ public Color4 ForStarDifficulty(double starDifficulty) => ColourUtils.SampleFrom
+        /// <summary>
+        /// Retrieves the colour for a given point in the star range.
+        /// </summary>
+        public Color4 ForStarDifficulty(double starDifficulty, bool showGrayOnZero = true)
```
I guess to offer something more productive: what is the flag for? where is it being turned off? *does it* need to be turned off?

is this just for the rainbow star rating slider or whatever? if it is can't you use `STAR_DIFFICULTY_SPECTRUM` directly there or something?

frenzibyte replied:
Tempted to actually make this show gray only on zero if the naming is misleading, I dunno about any other alternative names (`showGrayOnLessThanTenth`?). I initially had it flipped and named `continuous` but that's too vague than this.

bdach replied:
> Tempted to actually make this show gray only on zero

pray tell, how many beatmaps do you think this gray is going to be used on if you do this? do you recall that this was an explicit design decision? are you gonna update [web to do match your proposed behaviour too](https://github.com/ppy/osu-web/blob/c4809d06ba219d7ecba4f71665c7ffe741cc7405/resources/js/utils/beatmap-helper.ts#L84)?

maybe just remove the flag and inline a sample of `STAR_DIFFICULTY_SPECTRUM` in the one usage site I see (which is indeed the rainbow slider) that specifies `false` in your PR series?

frenzibyte replied:
sure, I just popped the wiki as my quick searching source for where the gray comes from and I couldn't find any trace of it, so I thought it's just a specific case for...beatmaps with unprocessed star rating maybe, I dunno.

this change of having a continuous spectrum is used in two places:
 - https://github.com/frenzibyte/osu/blob/c32feec37219b7ed9b0b99f81c3210605a8a52b8/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs#L53-L62
 - https://github.com/frenzibyte/osu/blob/c32feec37219b7ed9b0b99f81c3210605a8a52b8/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs#L151

the following diff can be applied to get rid of this and do things inline:
```diff
diff --git a/osu.Game/Graphics/OsuColour.cs b/osu.Game/Graphics/OsuColour.cs
index 0d9d22ffd3..dd5e19e167 100644
--- a/osu.Game/Graphics/OsuColour.cs
+++ b/osu.Game/Graphics/OsuColour.cs
@@ -22,7 +22,7 @@ public class OsuColour
 
         public static readonly (float, Color4)[] STAR_DIFFICULTY_SPECTRUM =
         {
-            (0.0f, Color4Extensions.FromHex("4290fb")),
+            (0.1f, Color4Extensions.FromHex("aaaaaa")),
             (0.1f, Color4Extensions.FromHex("4290fb")),
             (1.25f, Color4Extensions.FromHex("4fc0ff")),
             (2.0f, Color4Extensions.FromHex("4fffd5")),
@@ -40,13 +40,7 @@ public static readonly (float, Color4)[] STAR_DIFFICULTY_SPECTRUM =
         /// <summary>
         /// Retrieves the colour for a given point in the star range.
         /// </summary>
-        public Color4 ForStarDifficulty(double starDifficulty, bool showGrayOnZero = true)
-        {
-            if (showGrayOnZero && starDifficulty < 0.1f)
-                return Color4Extensions.FromHex("aaaaaa");
-
-            return ColourUtils.SampleFromLinearGradient(STAR_DIFFICULTY_SPECTRUM, (float)Math.Round(starDifficulty, 2, MidpointRounding.AwayFromZero));
-        }
+        public Color4 ForStarDifficulty(double starDifficulty) => ColourUtils.SampleFromLinearGradient(STAR_DIFFICULTY_SPECTRUM, (float)Math.Round(starDifficulty, 2, MidpointRounding.AwayFromZero));
 
         /// <summary>
         /// Retrieves the colour for a <see cref="ScoreRank"/>.
diff --git a/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs b/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs
index d22fce5a62..c873df2bfd 100644
--- a/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs
+++ b/osu.Game/Screens/SelectV2/FilterControl_DifficultyRangeSlider.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Extensions.Color4Extensions;
@@ -13,6 +14,7 @@
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Localisation;
 using osu.Game.Overlays;
+using osu.Game.Utils;
 using osuTK;
 using osuTK.Graphics;
 
@@ -26,6 +28,10 @@ public partial class DifficultyRangeSlider : ShearedRangeSlider
 
             private readonly LayoutValue drawSizeLayout = new LayoutValue(Invalidation.DrawSize);
 
+            private static readonly (float, Color4)[] spectrum = OsuColour.STAR_DIFFICULTY_SPECTRUM
+                                                                          .Skip(1)
+                                                                          .Prepend((0.0f, OsuColour.STAR_DIFFICULTY_SPECTRUM.ElementAt(1).Item2)).ToArray();
+
             public DifficultyRangeSlider()
                 : base("Star Rating")
             {
@@ -50,16 +56,15 @@ private void load(OverlayColourProvider colourProvider, OsuColour colours)
                         Shear = new Vector2(OsuGame.SHEAR, 0),
                         CornerRadius = 5f,
                         Masking = true,
-                        ChildrenEnumerable = OsuColour.STAR_DIFFICULTY_SPECTRUM
-                                                      .Zip(OsuColour.STAR_DIFFICULTY_SPECTRUM.Skip(1))
-                                                      .Select(p => new Box
-                                                      {
-                                                          RelativePositionAxes = Axes.X,
-                                                          X = p.First.Item1 / 10f,
-                                                          RelativeSizeAxes = Axes.Both,
-                                                          Width = (p.Second.Item1 - p.First.Item1) / 10f,
-                                                          Colour = ColourInfo.GradientHorizontal(p.First.Item2, p.Second.Item2),
-                                                      }),
+                        ChildrenEnumerable = spectrum.Zip(spectrum.Skip(1))
+                                                     .Select(p => new Box
+                                                     {
+                                                         RelativePositionAxes = Axes.X,
+                                                         X = p.First.Item1 / 10f,
+                                                         RelativeSizeAxes = Axes.Both,
+                                                         Width = (p.Second.Item1 - p.First.Item1) / 10f,
+                                                         Colour = ColourInfo.GradientHorizontal(p.First.Item2, p.Second.Item2),
+                                                     }),
                     },
                     borderContainer = new Container
                     {
@@ -148,7 +153,7 @@ protected override void LoadComplete()
 
                 protected override void UpdateDisplay(double value)
                 {
-                    Colour4 nubColour = colours.ForStarDifficulty(value, false);
+                    Colour4 nubColour = ColourUtils.SampleFromLinearGradient(spectrum, (float)Math.Round(value, 2, MidpointRounding.AwayFromZero));
                     nubColour = nubColour.Lighten(0.4f);
 
                     if (value >= 8.0)
```

@peppy will leave it to you as to whether you agree with spaceman this should be localised or not, the diff above makes it so.

frenzibyte replied:
have pushed necessary part myself, https://github.com/ppy/osu/pull/32764/commits/08c17bdf9ec27f9ee8ab35f9c9c2a3ae93b517f3.

bdach replied:
> sure, I just popped up the wiki as my quick searching source for where the gray comes from

you could have used, you know, git blame? you would have found https://github.com/ppy/osu/commit/9000d19c9eedca47292e4f283fdd0ab48a8dd9f3, then https://github.com/ppy/osu/pull/15597, then https://github.com/ppy/osu-web/pull/8317,  then https://github.com/ppy/osu-web/issues/8118#issuecomment-919939836 and https://github.com/ppy/osu-web/issues/7955

peppy commented on osu.Game/Graphics/OsuFont.cs:
```diff
@@ -16,14 +16,59 @@ public static class OsuFont
-        /// The default font.
+        /// Template font styles which should be preferred whenever possible for UI elements.
         /// </summary>
-        public static FontUsage Default => GetFont();
```
I've removed this change for now, good catch.

peppy replied:
Yeah, it'll be part of the mass change though – there's nothing actionable as long as we keep the batch changes out of this PR as it was removing a definition which is in use.

peppy commented:
I've dropped the last commit, will do that elsewhere (again, it was mostly for validation purposes).

frenzibyte commented on osu.Game/Graphics/OsuColour.cs:
```diff
@@ -403,6 +404,12 @@ public static Color4 ForegroundTextColourFor(Color4 backgroundColour)
         public readonly Color4 Orange3 = Color4Extensions.FromHex(@"cca633");
         public readonly Color4 Orange4 = Color4Extensions.FromHex(@"6b5c2e");
 
+        public readonly Color4 DarkOrange0 = Color4Extensions.FromHex(@"ffbb99");
+        public readonly Color4 DarkOrange1 = Color4Extensions.FromHex(@"ff9966");
+        public readonly Color4 DarkOrange2 = Color4Extensions.FromHex(@"eb7e47");
+        public readonly Color4 DarkOrange3 = Color4Extensions.FromHex(@"cc6633");
+        public readonly Color4 DarkOrange4 = Color4Extensions.FromHex(@"6b422e");
```
Used for fail-retry graph and user rating bar, computed from `OverlayColourProvider.Colour{0,4}` with 20 degrees hue:

![CleanShot 2025-04-11 at 07 14 05](https://github.com/user-attachments/assets/57a4a5af-3523-4067-aa0e-330e53931a41)
![CleanShot 2025-04-11 at 07 14 29](https://github.com/user-attachments/assets/65d032af-1dd2-48a2-978a-89efc47dab39)

I also vaguely remember osu-web having this set as well, but that's off-topic anyway.

bdach replied:
[raised with flyte](https://discord.com/channels/90072389919997952/892258907135836212/1360213077642973246)

frenzibyte committed:
Remove conditional in difficulty spectrum retrieval function

Wiil be handled locally instead using the diff in https://github.com/ppy/osu/pull/32764#discussion_r2039384833
